### DISCUSSION TITLE
How to extend dice example with sequence of dice values?

### DISCUSSION THREAD

**Discussion by svdoever** [microsoft/FluidFramework#8203] [Q&A]
For a project I would like to record a sequence of interactions for a certain user, but to understand the concept better I thought: let's take the FluidHelloWorld example (the dice rolling) and keep track of the sequence of dice values.

I would like to use the `SharedNumberSequence`, but there is not much documentation on its usage.

First step seems to be to extend the container schema and the `initialObjects`:

```javascript
const containerSchema = {
    initialObjects: { 
      diceMap: SharedMap,
      diceValues: SharedNumberSequence 
    }
};
```

Next step seems to be to initialize the `diceValues`, don't know how bu after searching the source code something like the following seems to be what is needed:

```javascript
container.initialObjects.diceMap.set(diceValueKey, 1);
container.initialObjects.diceValues = SharedNumberSequence.create(container.runtime, 'dice-values');
```

Is this correct? And what is the next step to add the initial "1" value to the `diceValues`? And how do I add the next value?
I expected I need to use the `insert` operation, e.g.:

```javascript
container.initialObjects.diceValues.insert(container.initialObjects.diceValues.getItemCount(), [1])
```

I assume this because in the implementation in the `ShareSequence<T>` class I see:

```typescript
/**
     * @param pos - The position to insert the items at.
     * @param items - The items to insert.
     * @param props - Optional. Properties to set on the inserted items.
     */
    public insert(pos: number, items: Serializable<T>[], props?: PropertySet) {
        const segment = new SubSequence<T>(items);
        if (props) {
            segment.addProperties(props);
        }
        const insertOp = this.client.insertSegmentLocal(pos, segment);
        if (insertOp) {
            this.submitSequenceMessage(insertOp);
        }
    }
```

But is this the right approach?

**Comment by svdoever** [comment]
In the mean time I got a lot further. See https://github.com/svdoever/FluidHelloWorld for an example implementation and some documentation on this.

**Comment by tylerbutler** [comment]
@svdoever You made impressive progress! I noticed in an earlier iteration of your code that you were checking for initialization of an initial object:

```ts
   if (!container.initialObjects.diceSequence) {
        log("Initial creation of dice sequence");
        try {
          container.initialObjects.diceSequence = SharedNumberSequence.create(
              container.runtime,
              "dice-sequence"
          );
        } catch (e) {
            log(`Error creating dice sequence: ${e}`);
        }
    }
```

You are correct that this code isn't correct or needed. [Initial objects are always initialized and available to you](https://fluidframework.com/docs/build/data-modeling/#defining-initial-objects) as part of loading the container. So you can just access the `diceSequence` immediately, no undefined check needed. I see that you removed this code so I assume your code works without it, but please let me know if you observed the `diceSequence` NOT being available. That would likely be a bug.

> And what is the next step to add the initial "1" value to the diceValues? And how do I add the next value?
> I expected I need to use the insert operation, e.g.:

Correct! `insert` is the method you want to use. I created an issue to track getting info about inherited methods/properties into child class reference docs: #8255. That should make it easier to find this in the future.

Also note that if you wanted to initialize some "default data" in an initial object, you can populate the data when the container is created, but before it is attached. There's more about this [in the fluidframework.com docs.](https://fluidframework.com/docs/build/containers/#attaching-a-container)

How was your experience working with Fluid overall? What was the most challenging? I'm excited that you're exploring the framework, so let me know how I can help!

**Reply by svdoever** [reply]
@tylerbutler I love the concept of Fluid framework, and it is great that the Azure Fluid Relay service is available now. It will fit exactly a missing piece of an application I'm building, and where I was researching how to share state between multiple clients. But there are a few things I just can't get my head around. What I need to do is create a sequence or recorded interaction per user. So I would like to have a `SharedMap` where as a key I have the id of the user, and as the value, I need to create a sequence. I can't model this upfront, because I don't no how many users there will be, and what their userId will be. 

I start by creating a container:

```javascript
    const containerSchema: ContainerSchema = {
        initialObjects: {
            userInteractionsMap: SharedMap
        },
    };
    const { container } = await client.createContainer(containerSchema);
    const containerId = await container.attach();
```

But next I want to have a function like:

```javascript
        addInteraction(userId: string, interaction: Interaction): void {
            const sharedUserInteractionsMap: ISharedMap = container.initialObjects.userInteractionsMap as SharedMap;
            const runtime = {} as IFluidDataStoreRuntime; // must be of type IFluidDataStoreRuntime, how do I get this?
            let userSpecificInteractionSequence = sharedUserInteractionsMap.get(userId);
            if (!userSpecificInteractionSequence) {
                const userSpecificInteractionSequence = SharedNumberSequence.create(
                    runtime as IFluidDataStoreRuntime,
                    "dice-sequence"
                );
                sharedUserInteractionsMap.set(userId, userSpecificInteractionSequence);
            }
            const userSpecificInteractionsCount = userSpecificInteractionSequence.getItemsCount();
            userSpecificInteractionSequence.insert(userSpecificInteractionsCount, interaction);
        }
```

But I have no clue, and can't find it in documentation, source code, or examples, how to get the `IFluidDataStoreRuntime`!

Another thing is also, how do I watch for changes in user-specific interaction sequences? Should I do something like:

```javascript
userSpecificInteractionSequence.on("sequenceDelta", (e) => {
    userSpecificSequenceDataChanged();
});
```

And how could I determine the userId whom sequence changed...

Any help that could point me in the right direction would be highly appreciated!

**Reply by tylerbutler** [reply]
> What I need to do is create a sequence or recorded interaction per user. So I would like to have a SharedMap where as a key I have the id of the user, and as the value, I need to create a sequence. I can't model this upfront, because I don't no how many users there will be, and what their userId will be.

This is a place where you want to use _dynamic objects._ These can be created dynamically at runtime. The way it works is that you declare the types you want to able to create at runtime, e.g. `SharedNumberSequence` in your case, and then call container.create to create instances of the dynamic object types. Note that this is always async. There are [more details in our documentation](https://fluidframework.com/docs/build/data-modeling/#dynamic-objects); I've pasted a code sample below. 

```ts
const schema = {
    /*...*/,
    dynamicObjectTypes: [ SharedCell, SharedMap ],
}

const { container, services } = await client.getContainer(id, schema);

const newCell = await container.create(SharedCell); // Create a new SharedCell
const newMap = await container.create(SharedMap); // Create a new SharedMap
```

> how do I watch for changes in user-specific interaction sequences?

You can listen for changes to individual sequences using the `sequenceDelta` event, as you guessed. Note that when you're storing handles to other DDSes in a SharedMap, you will _not_ get `valueChanged` events from SharedMap when the DDS in that map key is changed. If you want to track changes to the sequences, you need to set up event handlers for them.

> And how could I determine the userId whom sequence changed...

Do you need to infer this, or can you record who makes the change as part of your Fluid data?

**Reply by svdoever** [reply]
Thanks so much for your help. I now have some information to chew on. I keep you posted on my progress:-)
